오픈 마켓 [STEP1] 허황, 애플사이다

just1103 opened the pull request:
메이슨 안녕하세요. @myssun0325 
허황 @hwangjeha, 애플사이다 @just1103 입니다.
STEP1 완료하여 PR 드립니다. 네트워크 작업을 이번에 처음 해봐서 부족한 점이 많을텐데 잘 부탁드려요 :)

## STEP1 구현내용
### **OpenMarket**
- Extension
    - URLRequest+Extension : URL을 만들 때 HttpMethod까지 지정할 수 있도록 init() 추가
    - URLSession+Extension : 네트워크 없이 테스트를 진행하기 위한 URLSessionProtocol을 정의하고 URLSession 확장하여 채택함.
- Utility
    - JSONParser : JSON 데이터를 사용자 정의 타입으로 decode하기 위한 구조체 타입
        - struct인 이유 : enum + static method로 만들 수 있지만 static method는 항상 메모리에 올라가 있는 단점이 있어 struct로 구현함.
        - JSONDecoder().keyDecodingStrategy = .convertFromSnakeCase 지정하여 convertFromSnakeCase를 camelCase로 변환해서 데이터를 decode함.
    - NetworkDataTransfer : 서버와 데이터 통신을 하기 위한 구조체 타입
- Model
    - Product : 상품 상세 정보를 담는 모델 타입. Codable 프로토콜 채택
    - Currency : 상품 통화(나라별 화폐 단위) 정보를 담는 열거형 타입. Codable 프로토콜 채택
    - Image : 상품의 이미지 정보를 담는 모델 타입. Codable 프로토콜 채택
    - Vendor : 판매자 정보를 담는 모델 타입. Codable 프로토콜 채택
    - ProductPage : 상품 페이지 정보를 담는 모델 타입. Codable 프로토콜 채택
    - OpenMarketURL : OpenMarket 프로젝트에서 사용하는 URL를 관리하는 네임스페이스
    
### **OpenMarketTests**
- Mock
    - MockProduct.json : 가짜 상품 정보 json 데이터
    - MockProductPage.json : 가짜 상품 페이지 정보 json 데이터
    - MockURLSession : 네트워크 없이 테스트를 진행하기 위한 MockURLSession 타입. URLSessionProtocol 채택
- JSONParserTests :  JSONParser 타입을 테스트 하기 위한 테스트 코드. 가짜 json 데이터를 사용함.
- NetworkDataTransferTests : 서버와 통신을 하기 위한 테스트 코드. 네트워크 통신 없이 MockURLSession을 활용한 테스트도 진행

## 고민한 점
### 1. convertFromSnakeCase 및 CodingKeys 사용
convertFromSnakeCase를 사용하여 JSON Data의 SnakeCase로 된 parameter 이름을 Swift에서 CamelCase로 사용하도록 했습니다. 이때 convertFromSnakeCase가 처리할 수 없는 부분이 있었는데, 예를 들어 `page_no`는 Swift로 변환했을 때 `pageNo`이 아니라 `pageNumber`이 되어야 하므로 `CodingKey`를 함께 적용했습니다.
 
### 2. MockData로 테스트하기 위해 `JSON 파일` 사용
Swift 파일에 JSON String을 넣는 방법도 가능하지만, 향후 테스트할 대상 파일이 늘어날 것에 대비하여 JSON 파일 자체를 만들고 `Bundle.main.path`를 활용하여 데이터에 접근하는 방법을 사용했습니다.

### 3. 구현한 Model 타입 및 메서드
STEP1의 `서버 API 문서`에서 현재 확인 가능한 Model 타입만 구현했습니다. 마찬가지로 Post/Put/Delete 메서드도 STEP1에서는 Model 타입에 대한 정보가 불충분하여 GET 메서드만 구현했습니다. 

### 4. 탈출 클로저 및 Result<Data, Error>사용
기존에는 아래와 같이 `getHealthChecker` 메서드에서 반환타입을 설정하고, 비동기 처리 결과를 기다리도록 설정하기 위해 semaphore를 사용했습니다. 하지만 반환타입이 있는 이 방법으로는 `MockURLSession` 테스트가 불가하다고 판단해서 `탈출 클로저` 및 `Result<Data, Error>`를 활용하도록 수정하면서 semaphore를 삭제하고 `getHealthChecker` 메서드의 반환 타입을 변경했습니다.
```swift
// 수정 이전 - 반환타입 및 semaphore 사용
    private func getHealthChecker() -> Bool {
        guard let url = URL(string: "\(self.url)healthChecker") else {
            return false
        }
        var result: Bool = false
        
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "GET"
        
        let dataTask = URLSession.shared.dataTask(with: urlRequest) { _, response, _ in
            let successStatusCode = 200
            
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == successStatusCode else {
                      semaphore.signal()
                      return
                  }
            result = true
            semaphore.signal()
        }
        dataTask.resume()
        semaphore.wait()
        
        return result  
    }

// 수정 이후 - 탈출 클로저 사용
    func getHealthChecker(completionHandler: @escaping ((Result<Data, NetworkError>) -> Void)) {
        guard let urlRequest = URLRequest(url: OpenMarketURL.healthChecker, method: .get) else {
            return
        }
        loadData(request: urlRequest, completionHandler: completionHandler)
    }
```

## 궁금한 점
### 1. completionHandler의 error 타입
`let task = session.dataTask(with: request) { data, response, error in` 와 같이 completionHandler의 매개변수로 `error`가 포함되어 있는데, 서버 네트워킹이 성공하든, 실패하든 항상 `nil`인 것을 확인했습니다. 이 error 매개변수는 어떤 경우에 사용하는지 궁금합니다. 서버 관리자와 협의해야 하는 걸까요?

### 2. healthChecker 필요성
API 문서에서는 healthChecker를 통해 "OK" 데이터를 받아서 네트워킹이 성공했는지 확인하도록 하고 있습니다. 이렇게 서버와 데이터 통신할 때마다 매번 healthChecker로 확인해줄 필요가 있을까요? 현업에서 서버에 데이터를 보내기 전에 이러한 방식으로 확인 절차를 거치는지 궁금합니다.

## 조언을 듣고싶은 점
- 네이밍이 적절한지 궁금합니다. 특히 URLSession으로 데이터를 주고받는 부분에서 `NetWorkDataTransfer`라는 네이밍을 사용했는데, `APIManager` 등이 더 나을까요? (API가 너무 광범위한 용어라고 판단해서 "네트워크를 활용하여 데이터를 주고받는다"는 의미에서 NetWorkDataTransfer라고 네이밍했습니다.)
- 익스텐션/프로토콜을 사용하여 타입 의존성을 줄이고, 추상화를 하기 위해 고민했는데 이 부분이 적절한지 궁금합니다.

just1103 committed:
chore: MVC 폴더 추가 #1

Commit:
feat: Product 타입 추가 #1

- Currency 열거형 타입 추가

just1103 committed:
feat: Image 및 Vendor 타입 추가 #1

- Product 타입 Codable 프로토콜 채택

Commit:
feat: ProductPage 타입 추가 #1

just1103 committed:
feat: JSONParser 타입 및 decode 메서드 추가 #1

Commit:
feat: JSONParser 메서드 테스트 코드 추가 #1

- Mock Product JSON 문자열 생성

just1103 committed:
feat: MockProductPage 타입 및 테스트 코드 추가, SwiftLint Pod 추가 #1

Commit:
refactor: JSONParser decode() 매개변수 타입 Data?로 개선 #1

- 변경된 매개변수 타입에 맞게 테스트 코드 수정
- json 데이터 별도의 파일 생성

Commit:
docs: .gitignore Pods 폴더 설정

just1103 committed:
feat: NetworkDataTransfer 타입 및 getHealthChecker 메서드 추가 #1

- url, semaphore, isConnected 프로퍼티 추가
- isConnected 관련 테스트 코드 추가
- 테스트 코드 파일명 변경 (OpenMarketTests -> JSONParserTests)

Commit:
feat: getProductDetail 메서드 추가 #1

- Product vendors 프로퍼티 타입 변경 ([Vendor] -> Vendor)
- 테스트 코드 작성 test_getProductDetail가_정상작동_하는지

just1103 committed:
feat: getProductPage 메서드 추가 #1

- getProductPage 메서드 테스트 코드 추가

Commit:
refactor: 기존 NetworkDataTransfer 타입 리팩터링 #1

- dataTask 메서드 재사용 가능하게 개선
- getHealthChecker 매서드 개선

just1103 committed:
refactor: getProductDetail, getProductPage 메서드 리팩토링 #1

Commit:
feat: OpenMarketURL 열거형 추가 및 URLRequset 타입 확장 #1

- URLable 프로토콜 추가
- URLable 프로토콜, HttpMethod를 받는URLRequest 이니셜라이져 추가

just1103 committed:
feat: MockURLSession 클래스 및 테스트 코드 추가 #1

- URLSessionProtocol 프로토콜 추가
- URLSession 타입 익스텐션을 통해 URLSessionProtocol 채택

Commit:
refactor: NetworkDataTransfer 리팩터링 #1

- excaping Closure 사용 (매개변수로 completionHandler 받기 위함)
- 세마포어 제거
- 관련 테스트 코드 리팩터링

just1103 committed:
test: getProductDetail/getProductPage 및 MockURLSession 관련 테스트 코드 수정 #1

- MockURLSessionDataTask의 resume 메서드 관련 오류 수정
- MockURLSession 관련 성공/실패 케이스 테스트 코드 추가

Commit:
docs: STEP 1 README 작성 #1

just1103 committed:
test: OpenMarketTests Deployment Target 버전 변경 #1

- iOS 15 -> iOS 13.2

Commit:
refactor: URLProtocol 네이밍 개선 #1

- URLable -> URLProtocol

myssun0325 commented on OpenMarket/OpenMarket/Extension/URLRequest+Extension.swift:
```diff
@@ -0,0 +1,19 @@
+extension URLRequest {
+    init?(url: URLProtocol, method: HttpMethod) {
+        guard let url = url.url else {
+            return nil
```
이니셜라이져에서 url의 옵셔널 여부를 판단해서 Failable를 만드는 것과 외부에서 옵셔널 체크를 해서 Non Failable 이니셜라이져를 만드는 것에는 어떤 차이가 있다고 생각하나요?

hwangjeha replied:
보통은 NonFailable 이니셜라이져를 만드는게 보편적이고 안정적이라고 생각합니다.
하지만 이번 상황에서는 Failable 이니셜라이저를 사용하지 않으면 url을 반환할 때 강제추출 또는 옵셔널 바인딩을 해줘야했습니다.
저희는 각 메소드(getHealthChecker 등)에서 URLRequest를 바인딩해서 사용하기 때문에 Failable 이니셜라이저를 사용해도 된다고 판단했습니다.

myssun0325 commented on OpenMarket/OpenMarket/Extension/URLRequest+Extension.swift:
```diff
@@ -0,0 +1,19 @@
+    }
+}
+
+enum HttpMethod {
```
여기에 있는 타입이 이 파일에 있어야 하는 무엇인가요?

just1103 replied:
이 부분을 고민하다가 HttpMethod 열거형을 별도 파일로 분리하지 않았는데,
항상 URLRequest와 관련된 부분에서 HttpMethod 열거형을 쓰기 때문에 동일한 파일에 배치하는 것이 관리하기 편하다고 판단했기 때문입니다.
그래도 별도 파일로 분리하는 것이 좋을까요?

myssun0325 replied:
사실 파일 분리는 정해진 답이 없으니깐요 :). 
다만 지금 변경된 코드에서 보면 `APIProtocol`이 이 파일이 아니라 자연스레 Protocol폴더에 들어가야되는거 같고 그럼 HttpMethod는 APIProtocol과 붙어있으면 좋을지 URLRequest와 붙어있으면 좋을지 고민을 해보면 좋을거 같습니다 : )

just1103 replied:
APIProtocol을 별도 Protocol 폴더 내의 파일로 분리하고 HttpMethod 열거형도 해당 파일로 위치 변경했습니다.

myssun0325 commented on OpenMarket/OpenMarket/Extension/URLSession+Extension.swift:
```diff
@@ -0,0 +1,8 @@
+import Foundation
+
+protocol URLSessionProtocol {
```
Mock URLSession으로 테스트를 하기 위한 프로토콜 같아보이는군요 좋습니당 ☀️👍

myssun0325 commented on OpenMarket/OpenMarket/Model/Currency.swift:
```diff
@@ -0,0 +1,6 @@
+import Foundation
+
+enum Currency: String, Codable {
+    case koreanWon = "KRW"
```
enum을 사용할 때 케이스 별로 필요한 정보를 줘야할 때 rawValue를 쓰게 되면 한가지만 쓰게되는데요 만약에 case별로 여기서 사용한 "KRW", "USD"같은 rawValue를 말고 다른 것들도 써야하는 경우라면 어떻게 표현할 수 있을까요?

hwangjeha replied:
보통은 description 같은 연산 프로퍼티를 만들어 사용합니다.
JSON파일을 타입으로 decode할 때 rawValue를 이용하는 이니셜라이저를 사용하기 때문에 rawValue를 사용했습니다.

myssun0325 commented on OpenMarket/OpenMarket/Model/Image.swift:
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+struct Image: Codable {
```
Image라는 네이밍이 너무 포괄적인 네이밍이진 않을까 고민이 드네요~ 
지금은 Product라는 곳에서 쓰이니깐 Product이미지 일 수 있겠지만, 이미지가 여러종류고 여러종류의 이미지가 갖고 있는 정보가 다르다면 이미지간의 구분이 필요할 수 있을거 같네용 :)

just1103 replied:
말씀해주신 것처럼 여러 종류의 이미지가 사용될 수 있을 것 같아서 타입이름을 Image에서 ProductImage로 변경했습니다.

myssun0325 commented on OpenMarket/OpenMarket/Model/Product.swift:
```diff
@@ -0,0 +1,17 @@
+    let discountedPrice: Int
+    let stock: Int
+    let images: [Image]?
+    let vendors: Vendor?
```
vendors는 위에 images처럼 복수의 데이터를 나타낼거 같은 느낌이네요 😅

just1103 replied:
사실 처음에 vendors라는 네이밍만 보고 타입을 [Vendor]으로 지정해서 문제가 있었습니다. 오해를 막기 위해 CodingKeys를 활용하여 vendors를 vendor로 수정했습니다.

myssun0325 replied:
이건 서버쪽에 요청을 할만한 사항이네용 😭

myssun0325 commented on OpenMarket/OpenMarket/Model/ProductPage.swift:
```diff
@@ -0,0 +1,26 @@
+    let totalCount: Int
+    let offset: Int
+    let limit: Int
+    let pages: [Product]
```
pages라는 네이밍은 [Page]의 타입이 올거 같습니당

just1103 replied:
데이터를 확인해보니 특정 page의 Products를 나타내고 있어서 pages를 productsInPage로 변경했습니다.

myssun0325 replied:
여기는 Productpage로 구분이 되기 때문에 `let products: [Product]`로도 표현 가능하지 않을까합니다

hwangjeha replied:
자세하게 표현하려고 하다보니 productsInPage로 바꿨었는데 
타입 자체가 ProductPage라서 products로 변경해도 무방할 것 같습니다:)
의견 감사합니다:]

myssun0325 commented on OpenMarket/OpenMarket/Model/Vendor.swift:
```diff
@@ -0,0 +1,4 @@
+import Foundation
+
+struct Vendor: Codable {
```
이 녀석은 아직 안쓰이는것이여서 타입만 생성해둔걸까요? 
미리 예측을 해서 마련을 해두는것도 좋지만, 저는 예측보단 예측불가능한 미래의 요구사항을 반영할 수 있는 코드가 더 유연한 코드라고 생각합니다. 필요시에 추가/수정/삭제가 될 수 있는 유연한 코드라면 미래에 대비해서 지금 사용되지 않는 코드를 미리 작성할 필요가 있을까요?

hwangjeha replied:
야곰 아케데미에서 제공한 테스트 Product JSON 파일에는 vendors, images 키가 없습니다.
하지만, API 명세 페이지 상품 상세 조회Documentation -> Response 테이블을 보면
 vendors, images가 있어 vendor, image(productImage) 타입을 만들어줬습니다.

myssun0325 replied:
이번 스텝의 요구사항으로 Vendor가 필요한 값인가요? 
얘기해주신거처럼 지금 테스트에서 사용되지 않는 값인데 이번 스텝에서 필요한지에 대한 의문이었습니다 😅

myssun0325 replied:
API 명세 페이지에 있는 모든 걸 지금 구현해야하는가?에 대해서 저는 이번 스텝에서 요구하는 최소한의 것들에 대한 구현이나 받아오도록 하는게 좋다고 생각했어요

just1103 replied:
저희도 메이슨이 말씀해주신 것처럼 STEP1에 집중하는 것이 좋다고 판단했었습니다. 그런데 당시 API 명세의 상품조회 데이터를 열어보면 vendors, images에 데이터가 있어서 해당 타입이 필요할 것으로 예상했었는데, 다시 주석 처리해보니 데이터가 문제없이 받아지는 것을 확인했습니다. vendor, image를 STEP1에서 삭제했다가 다시 필요할 때 구현하도록 하겠습니다. 감사합니다 :)

myssun0325 commented on OpenMarket/OpenMarket/Utility/JSONParser.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+struct JSONParser<Item: Codable> {
+    func decode(from json: Data?) -> Item? {
```
이 메서드를 static으로 만드는것과 인스턴스의 메서드로 만드는 것은 어떤차이가 있을까요? (사용관점과 만드는 관점)

just1103 replied:
static으로 만들면 사용할 때 JSONParser 인스턴스가 없어도 해당 메서드를 사용할 수 있습니다. 하지만 static 메서드라서 메모리에 항상 올라가 있는 단점이 있습니다.
반면 인스턴스로 만들면 JSONParser 인스턴스가 필요합니다. 이 경우 원하는 곳에서 인스턴스를 생성하고, 사용이 끝나면 메모리에서 해제시키기 때문에 성능 측면에서 유리하다고 생각해서 인스턴스 메서드로 만들었습니다.

myssun0325 replied:
네 맞습니다. 하지만 다른 고려해볼 것으로 decode가 한번 쓰고 마는것이 아니라 계속 잦은 사용이 된다면 매번 인스턴스를 생성하고 해제하고 하는 비용을 고려해봤을 때도 인스턴스 메서드로 가는게 유리한가를 고민해 볼 수 있다 생각합니다. (물론 지금 이 부분이 성능에 큰 영향을 주는 포인트는 아닙니다 :) )

myssun0325 commented on OpenMarket/OpenMarket/Utility/NetworkDataTransfer.swift:
```diff
@@ -0,0 +1,59 @@
+import Foundation
+
+enum NetworkError: Error {
```
Error코드를 구분하셨군요 너무 좋습니당 👏

myssun0325 commented on OpenMarket/OpenMarket/Utility/NetworkDataTransfer.swift:
```diff
@@ -0,0 +1,59 @@
+    case unknownError
+}
+
+struct NetworkDataTransfer {
```
이 객체를 class와 struct 중 struct로 만드신 이유는 무엇인가요?

hwangjeha replied:
상속이나 참조를 사용하지 않기 때문에 애플에서 권장하는 struct를 사용했습니다. 

https://developer.apple.com/documentation/swift/choosing_between_structures_and_classes

myssun0325 commented on OpenMarket/OpenMarket/Utility/NetworkDataTransfer.swift:
```diff
@@ -0,0 +1,59 @@
+        self.session = session
+    }
+    
+    private func loadData(request: URLRequest, completionHandler: @escaping ((Result<Data, NetworkError>) -> Void)) {
```
개인적으로 이 메서드의 단위를 굉장히 잘 구분해서 만드신거 같습니다. 덕분에 다른 get하는 메서드들에서 활용이 잘되어지고 있군요 :) 👍

myssun0325 commented on OpenMarket/OpenMarketTests/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,48 @@
+                                              httpVersion: "2",
+                                              headerFields: nil)
+
+        let sessionDataTask = MockURLSessionDataTask()
```
큰 건 아니지만, MockURLSessionDataTask는 URLSessionDataTask를 상속받고있죠? 여기서 이 이니셜라이져는 deprecated되었다고 뜨는데 어떻게 해결하면 좋을까요?

just1103 replied:
애플에서 URLSessionDataTask를 직접 초기화하지 못하도록 deprecated 시킨 것을 파악했습니다. 
검색해보니 MockURLProtocol을 활용하여 해결하는 방법이 있었는데, (https://forums.raywenderlich.com/t/chapter-8-init-deprecated-in-ios-13/102050/8) 현업에서는 Alamofire를 활용하기 때문에 이 방법을 거의 사용하지 않는다고 알고 있습니다. 그래도 구현해보는 것이 좋을까요?

myssun0325 replied:
어떤 문제가 있고 어떤 방법이 있는지 아는것만으로도 충분합니다 👍

myssun0325 reviewed (commented):
수고하셨습니다. 🥳
전반적으로 코드가 깔끔하고 가독성이 좋아서 코드를 읽는데 큰 어려움이 없었고, 메서드의 단위도 잘 나누신거 같습니다. 👍 

### 궁금한 점 
1. completionHandler의 error 타입
말씀해주신 케이스가 서버에서 요청을 하고 서버에서 응답값으로 서버에러를 준 경우라는걸까요? 아니면 서버에 리퀘스트가 도달하지 못한 케이스일까요? 저희팀에서는 리퀘스트 자체가 실패났을 경우에 사용하고있습니다. 추가적으로 알라모아피라는 오픈소스도 참고해보시면 좋을듯합니당!! (Error 파트)

2. healthCheck 필요성
현업에서는 이런 API호출에 대해서 네트워크의 핑퐁을 따로 치고있진 않습니다. 어차피 서버의 상태가 안좋다면 미리 정해진 프로토콜로 에러가 내려오기 때문에 핑퐁을 치지 않아도 API호출만으로 상태를 확인할 수 있습니다. 응답이 없는 경우 클라이언트쪽에서 내려온 에러만으로 확인이 불가능한 경우 서버개발자와 커뮤니케이션을 하고있습니다.

3. 네이밍에 관하여
(NetWork보단 Network!) 저는 반대로 '네트워크'라는 개념이 API보다 더 아래 깔린 개념이라고 생각해요. 네트워크 환경을 통해 API를 사용하는거죠. 다른 오픈소스나 제가 사용하는 프로젝트에서의 레이어링을 한다면 현재 NetworkTransfer라는 단순히 네트워크의 기능만을 제공해주는 객체가 있고, API의 정보를 들고 있는 객체가 API의 정보들을 가지고 이 네트워크객체를 사용한다의 느낌이랄까요! 지금 NetworkDataTransfer라는 네이밍도 나쁘진 않아보입니다 하지만 흔히들 NetworkProvider, NetworkService 등의 네이밍이 보편화되어있는듯 하긴하네용 :). 
4. 프로토콜, 익스텐션에 의한 추상화
지금 코드에서 MockURLSession을 테스트하기위한 URLSession의 기능을 프로토콜로 추상화하신건 좋은거 같습니다. URLProtocol에 대한 생각은 조금 다릅니다. 만약에 URLProtocol을 추상화했다면 URLProtocol을 채택한 하나의 객체안에서 여러가지 케이스들을 늘려가는게 OpenMarketURL같은 구현체들이 여럿있고, 그 구현체마다 URLProtocol에서 요구하는 상황들을 갖도록하는게 유연하지 않을까 생각했어요. 이렇게 되면 API가 많아지면 실질적으로 채택한 구현체는 하나인데 그 객체만 커지는 그림이 그려져서요

just1103 committed:
refactor: JSON 매핑 모델의 프로퍼티명 수정 #2

Commit:
refactor: HttpMethod 열거형 연산 프로퍼티 추가 #2

just1103 committed:
refactor: decode 메서드의 반환타입 변경 #2

- JSONParserError 열거형 추가
- Item? -> Result<Item, JSONParserError>
- 관련 테스트 코드 수정

Commit:
refactor: Request URL 각각의 타입 생성 #2

- URLProtocol 채택
- BaseURLProtocol 추가
- 타입별 이니셜라이저에서 BaseURLProtocol 채택한 타입 의존성 주입

just1103 committed:
refactor: APIProtocol 추가하여 URLRequest 추상화 #2

- NetworkDataTransfer의 getRequest 메서드 추가하여 GET 메서드 추상화
- 관련 테스트 코드 수정

Commit:
refactor: JSONParserTests Bundle path 수정 #2

- NetworkDataTransfer 매서드 명 개선(getRequset -> request)

just1103 commented:
메이슨 안녕하세요. 어제 추상화 관련 디스코드로 피드백 주신 내용 반영해봤습니다.

## 1. BaseURLProtocol 및 APIProtocol 추가를 통한 추상화
아래와 같이 기존의 OpenMarketURL 열거형을 APIProtocol을 각각 채택한 구조체로 분리했습니다. 그리고 각 구조체의 이니셜라이저에서 baseURLProtocol을 채택한 타입을 의존성 주입하도록 했습니다. 
* 이렇게 분리해보니 get 관련 메서드들을 `request`메서드로 추상화하여 중복 코드를 제거할 수 있었습니다.
* 그리고 api 내부에 HTTP Method가 들어있고, URLRequest 내부에 body가 들어있어서 Get 뿐만 아니라 POST도 이 메서드로 가능하기 때문에 네이밍을 request로 지정했습니다. 추후 Post 메서드가 추가되면 APIProtocol에 body를 추가할 예정입니다.

```swift
struct OpenMarketBaseURL: BaseURLProtocol {
    var baseURL = "https://market-training.yagom-academy.kr/"
}

struct ProductDetailAPI: APIProtocol {
    var url: URL?
    var method: HttpMethod = .get
    
    init(id: Int, baseURL: BaseURLProtocol = OpenMarketBaseURL()) {
        self.url = URL(string: "\(baseURL.baseURL)api/products/\(id)")
    }
}
```

이 과정에서 HttpMethod 열거형을 사용하는 부분에서 `self.httpMethod = method.description` 가 어색하다고 느껴져서 아래와 같이 CustomStringConvertible을 채택하고, `self.httpMethod = "\(api.method)"` 형태로 사용하도록 수정했습니다.
```swift
extension URLRequest {
    init?(api: APIProtocol) {
        guard let url = api.url else {
            return nil
        }
        
        self.init(url: url)
        self.httpMethod = "\(api.method)"  // 수정한 부분
    }
}

enum HttpMethod: CustomStringConvertible {
    case get
    
    var description: String {
        switch self {
        case .get:
            return "GET"
        }
    }
}
```

## 2. 추가적으로 궁금한 점
URL 관련 하드코딩을 개선하기 위해 namespace 사용을 고민했는데, struct로 API들이 이미 분리된 상태라서 namespace를 사용하면 오히려 가독성이 떨어질 것 같아서 고민입니다. 메이슨은 어떻게 하실지 궁금합니다.

just1103 committed:
fix: request 메서드 Failure Result 구문 추가 #2

- URLRequest 옵셔널 바인딩 실패 시 Failure Result를 전달함
- 관련 테스트 코드 수정
- URL이 Nil이거나, 잘못된 URL인 경우에 대한 테스트 코드 추가

myssun0325 commented:
추가적으로 궁금한 점에 대해
- 오! 훨씬 깔끔하게 API 별로 나누셨군요! 👍 개인적인 의견을 드리자면, API 구현체마다 BaseURL이 필요한가?를 고민해볼거 같아요. 중복으로 계속 앞에 의미없이 넣어주는 값이 될 수도 있다고 생각해서요. 결국 API 내부에서 이 값은 뒤에 붙는 API url을 붙여서 최종 url을 만들어주고 이 최종 url을 request에 쓰게 되는데, request가 apiurl과 baseurl 조합하는 역할을 하거나 apiurl과 baseurl을 합쳐서 최종꼴을 만드는 프로토콜이나 객체를 만들어서 URLRequest에 넘겨주게 된다면 네임스페이스나 `"\(baseURL.baseURL)healthChecker")`와 같은 중복코드를 막을 수 있을거같아요. APIProtocol의 구현체들은 말그대로 서로 다른 API 문자열에 대해서만 책임을 갖고 값을 주는거죠.

Commit:
refactor: APIProtocol, HttpMethod 별도 파일로 분리 #2

just1103 committed:
refactor: ProductPage의 vendor, image 프로퍼티 삭제

- Vendor, ProductImage 타입 및 파일 삭제
- 해당 프로퍼티가 필요할 때 추가할 예정
- 프로퍼티명 수정 (productsInPage -> products)

hwangjeha commented:
baseURL을 APIProtocol에서 받거나, URLRequset 이니셜라이저에서 받게되면 `문자열 + URL`처리를 이니셜라이저에서 해야하고, ProductPageAPI 타입에서 사용중인 queryItem을 사용하는데 어려움이 있을 것 같습니다.
메이슨이 말씀해주신대로 별도의 타입이나 프로토콜을 사용하면 코드 중복을 줄일 수 있지만 아직 해결방안이 떠오르지 않아 조금 더 고민해보겠습니다😭 
꼼꼼한 피드백 감사합니다:]
